plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.rules.model;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.search.SearchType;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.core.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Simple abstract class that represent record stored in the Query Insight Framework
+ *
+ * @param <T> The value type associated with the record
+ * @opensearch.internal
+ */
+public abstract class SearchQueryRecord<T extends Number & Comparable<T>>
+    implements
+        Comparable<SearchQueryRecord<T>>,
+        Writeable,
+        ToXContentObject {
```

Comment:
This should be ToXContentFragment, and the subclasses should implement ToXContentObject

Reply:
Better yet, this one can implement `ToXContentObject`, emit the `startObject` and `endObject`, and can call an `abstract` method to emit the subclass fields.

Reply:
That's a good suggestion @msfroh! It also allows superclass to enforce same XContent format for all subclasses.

Reply:
Make sense!